File: DataStructAndAlgorithm/QuickSort.cpp
#include <stdio.h>

void Quick_sort(int* arr, int start, int end){
    if(start >= end) return;
    else if(end - start == 1){
        if(arr[start] > arr[end]){
            arr[start] ^= arr[end] ^= arr[start] ^= arr[end];
            return;
        }
    }
	int pivot, p, q;
	pivot = start;
	p = pivot + 1;
	q = end;
	while(p <= q){
		while(p <= end && arr[p] <= arr[pivot]) p++;
		while(q > start && arr[q] >= arr[pivot]) q--;
		if(p > q) arr[q] ^= arr[pivot] ^= arr[q] ^= arr[pivot];
		else arr[p] ^= arr[q] ^= arr[p] ^= arr[q];
	}
	Quick_sort(arr, start, q - 1);
	Quick_sort(arr, q + 1, end);
}

int main(){
	int arr[6] = {62, 98, 74, 39, 47, 53};
	Quick_sort(arr,0,5);
	for(int i=0;i<6;i++) printf("%d ",arr[i]);
	return 0;
}

File: DataStructAndAlgorithm/QuickSort.cs
public static class Extensions
{
    private static int DivideArray(int[] arr, int left, int right)
    {
        var pivotValue = arr[left];
        var leftIndex = left;
        var rightIndex = right;

        while (leftIndex < rightIndex)
        {
            while ((leftIndex <= right) && (arr[leftIndex] < pivotValue)) leftIndex++;
            while ((rightIndex >= left) && (arr[rightIndex] > pivotValue)) rightIndex--;

            if (leftIndex < rightIndex)
            {
                var temp = arr[leftIndex];
                arr[leftIndex] = arr[rightIndex];
                arr[rightIndex] = temp;

                if (arr[leftIndex] == arr[rightIndex]) rightIndex--;
            }
        }

        return rightIndex;
    }
    
    public static void QuickSort(this int[] arr, int left, int right)
    {
        if (left < right)
        {
            int pivot = DivideArray(arr, left, right);

            arr.QuickSort(left, pivot - 1);
            arr.QuickSort(pivot + 1, right);
        }
    }
}